aten/src/ATen/Quantizer.h

```diff
@@ -0,0 +1,205 @@
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct CAFFE2_API AffineQuantizer : public UniformQuantizer {
+  AffineQuantizer() {}
+  AffineQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~AffineQuantizer();
+};
+
+/**
+ * SymmetricQuantizer is similar to AffineQuantizer except that it
+ * does not have zero_point
+ *
+ * For quantize:
+ * Y = clamp(X * scale, min, max)
+ * For dequantize:
+ * X = Y / scale
+ */
+struct CAFFE2_API SymmetricQuantizer : public UniformQuantizer {
+  SymmetricQuantizer() {}
+  SymmetricQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~SymmetricQuantizer();
+
+};
+
+/**
+ * PerTensorSymmetricQuantizer stores a single scale number which is
+ * used for quantizing all the values in the given Tensor
+ */
+struct CAFFE2_API PerTensorSymmetricQuantizer: public SymmetricQuantizer {
+  PerTensorSymmetricQuantizer() {}
+  PerTensorSymmetricQuantizer(float scale) : SymmetricQuantizer(kPerTensorSymmetric), scale_(scale) {}
+  virtual ~PerTensorSymmetricQuantizer();
+  float scale_{1.0};
+};
+
+/**
+ * PerChannelSymmetricQuantizer stores a vector of scale number and
+ * applys symmetric quantization using different scales on each channel.
+ *
+ * Also note that per channel quantization is mostly applied to output channels of
+ * weights since per-input channel of weight quantization or per-channel quantization
+ * for activations can't be efficiently supported in most of processors since
+ * it requires each multiplication result within a single dot-product
+ * to have a different scale.
+ */
+struct CAFFE2_API PerChannelSymmetricQuantizer: public SymmetricQuantizer {
+  PerChannelSymmetricQuantizer() {}
+  PerChannelSymmetricQuantizer(std::vector<float> scales, std::vector<int64_t> axis): SymmetricQuantizer(kPerChannelSymmetric), scales_(scales), axis_(axis) {
+    AT_ASSERT(axis_.size() == 1);
+  }
+  virtual ~PerChannelSymmetricQuantizer();
+
+  std::vector<float> scales_;
+  std::vector<int64_t> axis_;
+};
+
+/**
+ * PerTensorAffineQuantizer stores a scale and a zero_point, which is used for
+ * all the values in the Tensor.
+ */
+struct CAFFE2_API PerTensorAffineQuantizer: public AffineQuantizer{
+  PerTensorAffineQuantizer(float scale, uint8_t zero_point): AffineQuantizer(kPerTensorAffine), scale_(scale), zero_point_(zero_point) {}
+  ~PerTensorAffineQuantizer();
+
+  virtual QTensor quantize(RealTensor tensor);
+  virtual RealTensor dequantize(QTensor tensor);
+
+  float scale() {
+    return scale_;
+  }
+
+  uint8_t zero_point() {
+    return zero_point_;
+  }
+
+  float scale_{1.0};
+  uint8_t zero_point_{0};
+};
+
+/**
+ * PerChannelAffineQuantizer is the same as PerTensorAffineQuantizer
+ * except that we have an independent scale and zero_point parameter
+ * for each channel.
+ */
+struct CAFFE2_API PerChannelAffineQuantizer: public AffineQuantizer {
```

Comment:
Will there be groupwise(symmetric|assymetric) quantizer as well?

Reply:
We can certainly add that if there is a need.

Reply:
Can we call it "PerAxis"? Because if I get it right, PerChannel and PerGroup just target different axis of the tensor. Similarly, we're going to have row-wise quantizer for embedding tables. So PerAxis with additional `axis` field seems appropriate.

Reply:
I thought per group is grouping channels into different groups and each group will have the same scale and zero_point, isn't it?